cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/BasicMapReduceContext.java

```diff
@@ -280,18 +285,38 @@ public void addOutput(String datasetName) {
 
   @Override
   public void addOutput(String datasetName, Map<String, String> arguments) {
-    // we can delay the instantiation of the Dataset to later, but for now, we still have to maintain backwards
-    // compatibility for the #setOutput(String, Dataset) method, so delaying the instantiation of this dataset will
-    // bring about code complexity without much benefit. Once #setOutput(String, Dataset) is removed, we can postpone
-    // this dataset instantiation
-    addOutput(datasetName, new DatasetOutputFormatProvider(datasetName, arguments,
-                                                           getDataset(datasetName, arguments, AccessType.WRITE),
-                                                           MapReduceBatchWritableOutputFormat.class));
+    addOutput(Output.ofDataset(datasetName, arguments));
   }
 
   @Override
   public void addOutput(String outputName, OutputFormatProvider outputFormatProvider) {
-    this.outputFormatProviders.put(outputName, outputFormatProvider);
+    addOutput(Output.of(outputName, outputFormatProvider));
+  }
+
+  @Override
+  public void addOutput(Output output) {
+    if (output instanceof Output.DatasetOutput) {
+      String datasetName = output.getName();
+      Map<String, String> arguments = ((Output.DatasetOutput) output).getArguments();
+
+      // we can delay the instantiation of the Dataset to later, but for now, we still have to maintain backwards
+      // compatibility for the #setOutput(String, Dataset) method, so delaying the instantiation of this dataset will
+      // bring about code complexity without much benefit. Once #setOutput(String, Dataset) is removed, we can postpone
+      // this dataset instantiation
+      DatasetOutputFormatProvider outputFormatProvider =
+        new DatasetOutputFormatProvider(datasetName, arguments,
+                                        getDataset(datasetName, arguments, AccessType.WRITE),
+                                        MapReduceBatchWritableOutputFormat.class);
+      this.outputFormatProviders.put(output.getAlias(), outputFormatProvider);
```

Comment:
If another Output with the same alias already exists, this will overwrite it? Would it be better to require that the aliases are unique?